{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "fc4a5eb4-2867-414b-ba86-5524521c22f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Isabel Jiah-Yih Liao\n",
    "# January 2024 \n",
    "# Synteny main \n",
    "\n",
    "##############\n",
    "### Notes: \n",
    "##############\n",
    "# Please ensure the correct intake files are in the correct directories prior to running. \n",
    "\n",
    "# If you would like to include an orthofinder run, create a directory under your root directory \n",
    "# called 'orthofinder_output' and place the results in the folder. If you are running orthofinder \n",
    "# within the script, do not include this file, as orthofinder will not run if this file already exists. \n",
    "\n",
    "# Expected files: \n",
    "#\n",
    "# root_directory (specified in settings) \n",
    "# ├── Synteny_main.ipynb\n",
    "# ├── dependencies\n",
    "# │   └── Synteny_functions.ipynb\n",
    "# ├── input_data\n",
    "# |   ├── gene_rows (directory for the .gtf files, doesn't need to be gene rows specifically)\n",
    "# |   │   ├── Bfl_gene_rows.gtf\n",
    "# |   │   ├── Cmu_gene_rows.gtf\n",
    "# |   │   ├── Mme_gene_rows.gtf\n",
    "# |   │   └── ...\n",
    "# |   ├── genomes\n",
    "# |   │   ├── Bfl.fna\n",
    "# |   │   ├── Cmu.fna\n",
    "# |   │   ├── Mme.fna\n",
    "# |   │   └── ...\n",
    "# |   └── proteomes\n",
    "# |       ├── Bfl.fasta\n",
    "# |       ├── Cmu.fasta\n",
    "# |       ├── Mme.fasta\n",
    "# |       └── ...\n",
    "# └── orthofinder_output (OPTIONAL INPUT: can also be run from this script) \n",
    "#     └── Results_MmmDD (can be any name) \n",
    "#         ├── Orthogroups\n",
    "#         │   ├── Orthogroups.tsv\n",
    "#         │   └── ...\n",
    "#         └── ... \n",
    "#\n",
    "# Make sure all input files have a backup copy to avoid losing information. \n",
    "\n",
    "# Resulting directory after running the program: \n",
    "# root_directory\n",
    "# ├── Synteny_main.ipynb\n",
    "# ├── dependencies\n",
    "# │   └── ...\n",
    "# ├── input_data\n",
    "# │   └── ...\n",
    "# └── run_name \n",
    "#     ├── output\n",
    "#     │   ├── Bfl_coordinates.gtf\n",
    "#     │   ├── Bfl_karyotype.gtf\n",
    "#     │   ├── Cmu_coordinates.gtf\n",
    "#     │   ├── Cmu_karyotype.gtf\n",
    "#     │   ├── Mme_coordinates.gtf\n",
    "#     │   └── Mme_karyotype.gtf\n",
    "#     └── run_files\n",
    "#         ├── orthofinder_output\n",
    "#         │   ├── Orthogroups\n",
    "#         │   │   └── ...\n",
    "#         │   └── ... \n",
    "#         └── run_proteomes\n",
    "#             ├── Bfl.fasta\n",
    "#             ├── Cmu.fasta\n",
    "#             ├── Mme.fasta\n",
    "#             └── ..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "22bd541f-e432-407b-ad1a-81073812b3a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "##########\n",
    "## Preparation \n",
    "##########\n",
    "\n",
    "# Import necessary dependencies\n",
    "import subprocess \n",
    "import os\n",
    "import re\n",
    "import pickle\n",
    "import pandas as pd\n",
    "from Bio import SeqIO \n",
    "%run dependencies/Synteny_functions.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e7ba2f09-ec07-4475-bd2a-aa3789450b77",
   "metadata": {},
   "outputs": [],
   "source": [
    "##########\n",
    "## Settings\n",
    "##########\n",
    "\n",
    "# Set up directories \n",
    "# Path to directory for running (directory containing input files)\n",
    "root_directory = './bryozoans' \n",
    "\n",
    "# List of codes for the respective species to include\n",
    "species_codes = [ 'Bfl', 'Bst', 'Cmu', 'Cpa', 'Mme', 'Wsu']\n",
    "run_name = 'bryozoa_full_test' \n",
    "\n",
    "# Location of your executable orthofinder \n",
    "orthofinder_path = \"/home/isabel/src/OrthoFinder/orthofinder\"\n",
    "threads = 150"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1780a335-0748-4058-b3ec-c31d8055011b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File check complete\n",
      "Directories built for run...\n",
      "No orthofinder results found .\n"
     ]
    }
   ],
   "source": [
    "##########\n",
    "## START\n",
    "##########\n",
    "\n",
    "# Create a Synteny object. Use .fasta for proteome file extensions rather than .fa. \n",
    "synteny = Synteny(root_directory, run_name, species_codes, proteome_ext = '.fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "396ffb7a-5072-4bb1-a00a-2418a8bd1727",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading karyotype files: this may take a moment...\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>OY755158.1</td>\n",
       "      <td>79350317</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>OY755159.1</td>\n",
       "      <td>77133738</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>OY755160.1</td>\n",
       "      <td>75932528</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OY755161.1</td>\n",
       "      <td>67125504</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>OY755162.1</td>\n",
       "      <td>66617907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>OY755163.1</td>\n",
       "      <td>66545008</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>OY755164.1</td>\n",
       "      <td>66307180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>OY755165.1</td>\n",
       "      <td>65553477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>OY755166.1</td>\n",
       "      <td>64042602</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>OY755167.1</td>\n",
       "      <td>61084640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>OY755168.1</td>\n",
       "      <td>46061466</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Chromosome    Length\n",
       "0   OY755158.1  79350317\n",
       "1   OY755159.1  77133738\n",
       "2   OY755160.1  75932528\n",
       "3   OY755161.1  67125504\n",
       "4   OY755162.1  66617907\n",
       "5   OY755163.1  66545008\n",
       "6   OY755164.1  66307180\n",
       "7   OY755165.1  65553477\n",
       "8   OY755166.1  64042602\n",
       "9   OY755167.1  61084640\n",
       "10  OY755168.1  46061466"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "##########\n",
    "## Creating the karyotype files\n",
    "##########\n",
    "\n",
    "# Read the genome file to find the chromosomes and their associated lengths. \n",
    "# Specify the number of chromosomes n in each species. The longest n scaffolds \n",
    "# in each genome will be taken as the chromosomes. \n",
    "\n",
    "# Tip: If you are uncertain about the number of chromosomes, set the karyotype values \n",
    "# higher and look at how the scaffolds are named. \n",
    "\n",
    "chromosomes = {'Bfl': 19, 'Bst': 11, 'Cmu': 8, 'Cpa': 12, 'Mme': 11, 'Wsu': 11 }\n",
    "synteny.build_karyotype(chromosomes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b799423b-608b-42ce-a97f-2417b6f3c93b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>species</th>\n",
       "      <th>size</th>\n",
       "      <th>color</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>OY755158.1</td>\n",
       "      <td>1</td>\n",
       "      <td>79350317</td>\n",
       "      <td>Wsu</td>\n",
       "      <td>12</td>\n",
       "      <td>25252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>OY755159.1</td>\n",
       "      <td>1</td>\n",
       "      <td>77133738</td>\n",
       "      <td>Wsu</td>\n",
       "      <td>12</td>\n",
       "      <td>25252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>OY755160.1</td>\n",
       "      <td>1</td>\n",
       "      <td>75932528</td>\n",
       "      <td>Wsu</td>\n",
       "      <td>12</td>\n",
       "      <td>25252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OY755161.1</td>\n",
       "      <td>1</td>\n",
       "      <td>67125504</td>\n",
       "      <td>Wsu</td>\n",
       "      <td>12</td>\n",
       "      <td>25252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>OY755162.1</td>\n",
       "      <td>1</td>\n",
       "      <td>66617907</td>\n",
       "      <td>Wsu</td>\n",
       "      <td>12</td>\n",
       "      <td>25252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>OY755163.1</td>\n",
       "      <td>1</td>\n",
       "      <td>66545008</td>\n",
       "      <td>Wsu</td>\n",
       "      <td>12</td>\n",
       "      <td>25252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>OY755164.1</td>\n",
       "      <td>1</td>\n",
       "      <td>66307180</td>\n",
       "      <td>Wsu</td>\n",
       "      <td>12</td>\n",
       "      <td>25252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>OY755165.1</td>\n",
       "      <td>1</td>\n",
       "      <td>65553477</td>\n",
       "      <td>Wsu</td>\n",
       "      <td>12</td>\n",
       "      <td>25252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>OY755166.1</td>\n",
       "      <td>1</td>\n",
       "      <td>64042602</td>\n",
       "      <td>Wsu</td>\n",
       "      <td>12</td>\n",
       "      <td>25252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>OY755167.1</td>\n",
       "      <td>1</td>\n",
       "      <td>61084640</td>\n",
       "      <td>Wsu</td>\n",
       "      <td>12</td>\n",
       "      <td>25252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>OY755168.1</td>\n",
       "      <td>1</td>\n",
       "      <td>46061466</td>\n",
       "      <td>Wsu</td>\n",
       "      <td>12</td>\n",
       "      <td>25252</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Chr Start       End species size  color\n",
       "0   OY755158.1     1  79350317     Wsu   12  25252\n",
       "1   OY755159.1     1  77133738     Wsu   12  25252\n",
       "2   OY755160.1     1  75932528     Wsu   12  25252\n",
       "3   OY755161.1     1  67125504     Wsu   12  25252\n",
       "4   OY755162.1     1  66617907     Wsu   12  25252\n",
       "5   OY755163.1     1  66545008     Wsu   12  25252\n",
       "6   OY755164.1     1  66307180     Wsu   12  25252\n",
       "7   OY755165.1     1  65553477     Wsu   12  25252\n",
       "8   OY755166.1     1  64042602     Wsu   12  25252\n",
       "9   OY755167.1     1  61084640     Wsu   12  25252\n",
       "10  OY755168.1     1  46061466     Wsu   12  25252"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# To format the karyotype file, pass a list containing the vales of each column. \n",
    "# 'Chromosome', 'Length', and 'SPECIES' are all keywords allowing their associated data to be pulled. \n",
    "# Any non-keywords in the 'columns' list will yield a column containing that value. \n",
    "# Optionally, pass a 'labels' variable to relabel the column names. \n",
    "columns = ['Chromosome', '1', 'Length', 'SPECIES', '12', '25252']\n",
    "labels = ['Chr', 'Start', 'End', 'species', 'size', 'color'] \n",
    "synteny.clean_karyotype(columns, labels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "cf0afd2d-7ba5-4dd0-b8ec-f4de7570aae8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write the karyotype files to the output folder. \n",
    "synteny.write_karyotype()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "84d0d4b0-18c5-4e89-86da-9672bd3ab896",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading proteomes...\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[SeqRecord(seq=Seq('MVTDPLRALLGKGADFKWEQRQEKAFQEVKQLVLSEHPTQVQIPVNRQNCVRNE...NYX'), id='g312.t1', name='g312.t1', description='g312.t1', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('MSRKDNAVDGDKVSKYEQTLRRASNRPEVILCSFYPSTRVPAAQYGIIGERDTR...LV*'), id='g313.t1', name='g313.t1', description='g313.t1', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('MRLGWSTYKTDANTALHPYWDEKGHFPERGTESVVDKEKEKRERTMMNYNKRHR...LY*'), id='g314.t1', name='g314.t1', description='g314.t1', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('MGKEGLEHYIERLWARYDGDEASDEASDEASDEASDEASDEASDEASDEASDEA...GL*'), id='g315.t1', name='g315.t1', description='g315.t1', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('MERGVLQGFVDYVYGLVTAQRVSEVHVVAACSEAWTANPKEYICIQENHLGVPV...DG*'), id='g316.t1', name='g316.t1', description='g316.t1', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('MNYPFQYYALKVLRRPEMKLYQVNKESQRLERWLRASTGSLLKASGVKLYQVNK...KD*'), id='g317.t1', name='g317.t1', description='g317.t1', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('MYEYQDIVSLINHYPGHESLWYYRKWLVYQLSLQQESSCRLTVEGELELEGEDD...DD*'), id='g318.t1', name='g318.t1', description='g318.t1', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('MSGAPAPDCIGTPVPPDTTVTPAPMAAAGTPVMSDATSTSVLAAIVGTPVPFDM...RH*'), id='g319.t1', name='g319.t1', description='g319.t1', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('MIQGLRVNMENRPAINAEAILKTAELSKEDHPRVDRLFREPKYVDNVGELVENE...VL*'), id='g320.t1', name='g320.t1', description='g320.t1', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('MSSASGATAEAILEVTHPVRLKVELVTLVKDEGISRARNTVQTAHQHRKTLHDK...PW*'), id='g321.t1', name='g321.t1', description='g321.t1', dbxrefs=[])]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read the proteomes into Python \n",
    "synteny.read_proteomes()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f5fdf80a-fc7d-4059-8dd8-567367d53863",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[SeqRecord(seq=Seq('MMLARNPMPSRTRSLITAGDTTSTADYFASHNTASNRTLADPPSHVGLPPLKSQ...IVG'), id='BL09372', name='7948_Ea_BL09372', description='7948_Ea_BL09372', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('MATRLDRLFTLLDTGSTPVIRKSAALQIGQVQKLHPHELHNLLAKVLTFLRSDN...SLT'), id='BL24099', name='8096_H_BL24099', description='8096_H_BL24099', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('MRTSDHRRLIHRDVWPFSGPFLFGRCRPAGPGKGGLEERAPPPRGDAQGVGHPP...DKH'), id='BL02556', name='9923_F_BL02556', description='9923_F_BL02556', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('MTLTTGLSQPFRRLDKYPTLLKELDRHLEEGHPDRYDVQQAIPVYKNIAERKER...TNL'), id='BL13883', name='9315_N_BL13883', description='9315_N_BL13883', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('MNRDQVINFAPGPAKLPQEVLEQAQKDMLSYNNLGISVMEMSHRSSDFSKIINN...TNQ'), id='BL16740', name='3413_C2_BL16740', description='3413_C2_BL16740', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('MSGLHSKHGFKVEALKPKAGGKAVHGSAIPLILPNDLEYNVSTGAVEKVQGARR...LKH'), id='BL01007', name='5160_A1a_BL01007', description='5160_A1a_BL01007', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('MGANASTGGPGEKIPNPQYRTSALLHRYSSVDTPLDPDVGHGIPHTLAASRHHS...SYS'), id='BL14725', name='9197_A1b_BL14725', description='9197_A1b_BL14725', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('MSGPEKGVDTDPDYAVVCGFLRRYGEALGLPPMSYAELQRALEDTTRLNPDLIL...EDD'), id='BL25804', name='4401_N_BL25804', description='4401_N_BL25804', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('MKLTDGTLRSFKVAKVFRENSDRINSLDFAPSGETLISSSDDDSIVIYDCQEGK...TDD'), id='BL28013', name='9658_Eb_BL28013', description='9658_Eb_BL28013', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('MPTKVPFKVIQAASQDDGFSAKELETHAPTVNGWVSSRFCLYPQELVLKLNERV...RGK'), id='BL04804', name='4866_P_BL04804', description='4866_P_BL04804', dbxrefs=[])]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# We are using Bfl for ALGs, but we want the gene names to still fit the format. This is a tool \n",
    "# which renames genes to a substring of the current name. This allows the gene ids to match the \n",
    "# annotations later on. However, it only needs to be run on Bfl or other similarily formatted \n",
    "# proteomes. \n",
    "\n",
    "# Given longer ids with some kind of a pattern, take only a portion of the id as the gene name. \n",
    "# For instance, if the id is XXX_XXXX_GENEID we want to keep only the third column, using '_' as \n",
    "# a delimiter. \n",
    "# If no delimiter is specified, '_' is assumed. \n",
    "synteny.proteome_id_trim('Bfl', position = 3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f166ae43-a103-4947-a559-ef94d81efccd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Proteome gene names modified for Bfl\n",
      "Proteome gene names modified for Bst\n",
      "Proteome gene names modified for Cmu\n",
      "Proteome gene names modified for Cpa\n",
      "Proteome gene names modified for Mme\n",
      "Proteome gene names modified for Wsu\n",
      "Bfl proteomes successfully written to ./bryozoans/bryozoa_full_test/run_files/run_proteomes/Bfl.fa.\n",
      "Bst proteomes successfully written to ./bryozoans/bryozoa_full_test/run_files/run_proteomes/Bst.fa.\n",
      "Cmu proteomes successfully written to ./bryozoans/bryozoa_full_test/run_files/run_proteomes/Cmu.fa.\n",
      "Cpa proteomes successfully written to ./bryozoans/bryozoa_full_test/run_files/run_proteomes/Cpa.fa.\n",
      "Mme proteomes successfully written to ./bryozoans/bryozoa_full_test/run_files/run_proteomes/Mme.fa.\n",
      "Wsu proteomes successfully written to ./bryozoans/bryozoa_full_test/run_files/run_proteomes/Wsu.fa.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[SeqRecord(seq=Seq('MVTDPLRALLGKGADFKWEQRQEKAFQEVKQLVLSEHPTQVQIPVNRQNCVRNE...NYX'), id='Wsu|g312.t1', name='g312.t1', description='', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('MSRKDNAVDGDKVSKYEQTLRRASNRPEVILCSFYPSTRVPAAQYGIIGERDTR...LV*'), id='Wsu|g313.t1', name='g313.t1', description='', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('MRLGWSTYKTDANTALHPYWDEKGHFPERGTESVVDKEKEKRERTMMNYNKRHR...LY*'), id='Wsu|g314.t1', name='g314.t1', description='', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('MGKEGLEHYIERLWARYDGDEASDEASDEASDEASDEASDEASDEASDEASDEA...GL*'), id='Wsu|g315.t1', name='g315.t1', description='', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('MERGVLQGFVDYVYGLVTAQRVSEVHVVAACSEAWTANPKEYICIQENHLGVPV...DG*'), id='Wsu|g316.t1', name='g316.t1', description='', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('MNYPFQYYALKVLRRPEMKLYQVNKESQRLERWLRASTGSLLKASGVKLYQVNK...KD*'), id='Wsu|g317.t1', name='g317.t1', description='', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('MYEYQDIVSLINHYPGHESLWYYRKWLVYQLSLQQESSCRLTVEGELELEGEDD...DD*'), id='Wsu|g318.t1', name='g318.t1', description='', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('MSGAPAPDCIGTPVPPDTTVTPAPMAAAGTPVMSDATSTSVLAAIVGTPVPFDM...RH*'), id='Wsu|g319.t1', name='g319.t1', description='', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('MIQGLRVNMENRPAINAEAILKTAELSKEDHPRVDRLFREPKYVDNVGELVENE...VL*'), id='Wsu|g320.t1', name='g320.t1', description='', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('MSSASGATAEAILEVTHPVRLKVELVTLVKDEGISRARNTVQTAHQHRKTLHDK...PW*'), id='Wsu|g321.t1', name='g321.t1', description='', dbxrefs=[])]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Modify select headers to add species code to gene names and remove description information. \n",
    "# This would yield more precise proteomes to feed into orthofinder. \n",
    "# Run this with either a list containing a subset of species to modify, or run without \n",
    "# any parameters to modify all proteomes in species_codes. \n",
    "synteny.proteome_add_species()\n",
    "synteny.write_proteomes()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "8ac0252e-e3cf-435c-a88a-d38b26afd841",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['nohup', '/home/isabel/src/OrthoFinder/orthofinder', '-f', './bryozoans/bryozoa_full_test/run_files/run_proteomes', '-o', './bryozoans/bryozoa_full_test/run_files/orthofinder_output', '-t', '150']\n",
      "\n",
      "OrthoFinder version 2.5.4 Copyright (C) 2014 David Emms\n",
      "\n",
      "2024-01-19 09:28:36 : Starting OrthoFinder 2.5.4\n",
      "150 thread(s) for highly parallel tasks (BLAST searches etc.)\n",
      "16 thread(s) for OrthoFinder algorithm\n",
      "\n",
      "Checking required programs are installed\n",
      "----------------------------------------\n",
      "Test can run \"mcl -h\" - ok\n",
      "Test can run \"fastme -i ./bryozoans/bryozoa_full_test/run_files/orthofinder_output/Results_Jan19/WorkingDirectory/SimpleTest.phy -o ./bryozoans/bryozoa_full_test/run_files/orthofinder_output/Results_Jan19/WorkingDirectory/SimpleTest.tre\" - ok\n",
      "\n",
      "Dividing up work for BLAST for parallel processing\n",
      "--------------------------------------------------\n",
      "2024-01-19 09:28:37 : Creating diamond database 1 of 6\n",
      "2024-01-19 09:28:37 : Creating diamond database 2 of 6\n",
      "2024-01-19 09:28:37 : Creating diamond database 3 of 6\n",
      "2024-01-19 09:28:38 : Creating diamond database 4 of 6\n",
      "2024-01-19 09:28:38 : Creating diamond database 5 of 6\n",
      "2024-01-19 09:28:38 : Creating diamond database 6 of 6\n",
      "\n",
      "Running diamond all-versus-all\n",
      "------------------------------\n",
      "Using 150 thread(s)\n",
      "2024-01-19 09:28:38 : This may take some time....\n",
      "2024-01-19 09:31:21 : Done all-versus-all sequence search\n",
      "\n",
      "Running OrthoFinder algorithm\n",
      "-----------------------------\n",
      "2024-01-19 09:31:21 : Initial processing of each species\n",
      "2024-01-19 09:31:23 : Initial processing of species 0 complete\n",
      "2024-01-19 09:31:30 : Initial processing of species 5 complete\n",
      "2024-01-19 09:31:30 : Initial processing of species 2 complete\n",
      "2024-01-19 09:31:31 : Initial processing of species 1 complete\n",
      "2024-01-19 09:31:34 : Initial processing of species 4 complete\n",
      "2024-01-19 09:31:35 : Initial processing of species 3 complete\n",
      "2024-01-19 09:31:39 : Connected putative homologues\n",
      "2024-01-19 09:31:40 : Written final scores for species 0 to graph file\n",
      "2024-01-19 09:31:42 : Written final scores for species 1 to graph file\n",
      "2024-01-19 09:31:42 : Written final scores for species 5 to graph file\n",
      "2024-01-19 09:31:42 : Written final scores for species 4 to graph file\n",
      "2024-01-19 09:31:43 : Written final scores for species 2 to graph file\n",
      "2024-01-19 09:31:43 : Written final scores for species 3 to graph file\n",
      "\n",
      "WARNING: program called by OrthoFinder produced output to stderr\n",
      "\n",
      "Command: mcl ./bryozoans/bryozoa_full_test/run_files/orthofinder_output/Results_Jan19/WorkingDirectory/OrthoFinder_graph.txt -I 1.5 -o ./bryozoans/bryozoa_full_test/run_files/orthofinder_output/Results_Jan19/WorkingDirectory/clusters_OrthoFinder_I1.5.txt -te 16 -V all\n",
      "\n",
      "stdout\n",
      "------\n",
      "\n",
      "stderr\n",
      "------\n",
      "[mcl] cut <1> instances of overlap\n",
      "\n",
      "2024-01-19 09:32:02 : Ran MCL\n",
      "\n",
      "Writing orthogroups to file\n",
      "---------------------------\n",
      "OrthoFinder assigned 99274 genes (76.3% of total) to 16861 orthogroups. Fifty percent of all genes were in orthogroups with 5 or more genes (G50 was 5) and were contained in the largest 7284 orthogroups (O50 was 7284). There were 1976 orthogroups with all species present and 1252 of these consisted entirely of single-copy genes.\n",
      "\n",
      "2024-01-19 09:32:06 : Done orthogroups\n",
      "\n",
      "Analysing Orthogroups\n",
      "=====================\n",
      "\n",
      "Calculating gene distances\n",
      "--------------------------\n",
      "2024-01-19 09:32:24 : Done\n",
      "2024-01-19 09:32:25 : Done 0 of 12533\n",
      "2024-01-19 09:32:26 : Done 1000 of 12533\n",
      "2024-01-19 09:32:26 : Done 2000 of 12533\n",
      "2024-01-19 09:32:26 : Done 3000 of 12533\n",
      "2024-01-19 09:32:26 : Done 4000 of 12533\n",
      "2024-01-19 09:32:26 : Done 5000 of 12533\n",
      "2024-01-19 09:32:27 : Done 6000 of 12533\n",
      "2024-01-19 09:32:27 : Done 7000 of 12533\n",
      "2024-01-19 09:32:27 : Done 8000 of 12533\n",
      "2024-01-19 09:32:28 : Done 9000 of 12533\n",
      "2024-01-19 09:32:28 : Done 10000 of 12533\n",
      "2024-01-19 09:32:28 : Done 11000 of 12533\n",
      "2024-01-19 09:32:28 : Done 12000 of 12533\n",
      "\n",
      "Inferring gene and species trees\n",
      "--------------------------------\n",
      "\n",
      "1976 trees had all species present and will be used by STAG to infer the species tree\n",
      "\n",
      "Best outgroup(s) for species tree\n",
      "---------------------------------\n",
      "2024-01-19 09:33:21 : Starting STRIDE\n",
      "2024-01-19 09:33:25 : Done STRIDE\n",
      "Observed 128 well-supported, non-terminal duplications. 127 support the best root and 1 contradict it.\n",
      "Best outgroup for species tree:\n",
      "  Bfl\n",
      "\n",
      "Reconciling gene trees and species tree\n",
      "---------------------------------------\n",
      "Outgroup: Bfl\n",
      "2024-01-19 09:33:25 : Starting Recon and orthologues\n",
      "2024-01-19 09:33:25 : Starting OF Orthologues\n",
      "2024-01-19 09:33:26 : Done 0 of 12533\n",
      "2024-01-19 09:33:26 : Done 1000 of 12533\n",
      "2024-01-19 09:33:26 : Done 2000 of 12533\n",
      "2024-01-19 09:33:27 : Done 3000 of 12533\n",
      "2024-01-19 09:33:27 : Done 4000 of 12533\n",
      "2024-01-19 09:33:27 : Done 5000 of 12533\n",
      "2024-01-19 09:33:27 : Done 6000 of 12533\n",
      "2024-01-19 09:33:27 : Done 7000 of 12533\n",
      "2024-01-19 09:33:27 : Done 8000 of 12533\n",
      "2024-01-19 09:33:27 : Done 9000 of 12533\n",
      "2024-01-19 09:33:28 : Done 10000 of 12533\n",
      "2024-01-19 09:33:28 : Done 11000 of 12533\n",
      "2024-01-19 09:33:28 : Done 12000 of 12533\n",
      "2024-01-19 09:33:29 : Done OF Orthologues\n",
      "\n",
      "Writing results files\n",
      "=====================\n",
      "2024-01-19 09:33:30 : Done orthologues\n",
      "\n",
      "Results:\n",
      "    bryozoans/bryozoa_full_test/run_files/orthofinder_output/Results_Jan19/\n",
      "\n",
      "CITATION:\n",
      " When publishing work that uses OrthoFinder please cite:\n",
      " Emms D.M. & Kelly S. (2019), Genome Biology 20:238\n",
      "\n",
      " If you use the species tree in your work then please also cite:\n",
      " Emms D.M. & Kelly S. (2017), MBE 34(12): 3267-3278\n",
      " Emms D.M. & Kelly S. (2018), bioRxiv https://doi.org/10.1101/267914\n"
     ]
    }
   ],
   "source": [
    "##########\n",
    "## Running OrthoFinder\n",
    "##########\n",
    "# This only needs to be done once per set of species so keep it commented out to avoid running \n",
    "# it accidentally\n",
    "synteny.run_orthofinder(orthofinder_path, threads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "7f137104-e55c-4655-a144-1053a64efa73",
   "metadata": {},
   "outputs": [],
   "source": [
    "synteny.incorporate_orthofinder()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "e40781ec-2207-4f08-85c0-dd4e1079155b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Orthogroup</th>\n",
       "      <th>Bfl</th>\n",
       "      <th>Bst</th>\n",
       "      <th>Cmu</th>\n",
       "      <th>Cpa</th>\n",
       "      <th>Mme</th>\n",
       "      <th>Wsu</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3794</th>\n",
       "      <td>OG0003794</td>\n",
       "      <td>Bfl|BL24099</td>\n",
       "      <td>Bst|g3299.t1</td>\n",
       "      <td>Cmu|g20770.t1</td>\n",
       "      <td>Cpa|g29337.t1</td>\n",
       "      <td>Mme|g1166.t1</td>\n",
       "      <td>Wsu|g824.t1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3795</th>\n",
       "      <td>OG0003795</td>\n",
       "      <td>Bfl|BL02556</td>\n",
       "      <td>Bst|g18397.t1</td>\n",
       "      <td>Cmu|g17124.t1</td>\n",
       "      <td>Cpa|g9366.t1</td>\n",
       "      <td>Mme|g13681.t1</td>\n",
       "      <td>Wsu|g16010.t1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3796</th>\n",
       "      <td>OG0003796</td>\n",
       "      <td>Bfl|BL13883</td>\n",
       "      <td>Bst|g12973.t2</td>\n",
       "      <td>Cmu|g25307.t1</td>\n",
       "      <td>Cpa|g23633.t1</td>\n",
       "      <td>Mme|g25764.t2</td>\n",
       "      <td>Wsu|g20045.t1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3797</th>\n",
       "      <td>OG0003797</td>\n",
       "      <td>Bfl|BL16740</td>\n",
       "      <td>Bst|g16271.t1</td>\n",
       "      <td>Cmu|g19511.t1</td>\n",
       "      <td>Cpa|g6068.t1</td>\n",
       "      <td>Mme|g18655.t1</td>\n",
       "      <td>Wsu|g11821.t1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3798</th>\n",
       "      <td>OG0003798</td>\n",
       "      <td>Bfl|BL14725</td>\n",
       "      <td>Bst|g11405.t1</td>\n",
       "      <td>Cmu|g18040.t1</td>\n",
       "      <td>Cpa|g2841.t1</td>\n",
       "      <td>Mme|g16154.t1</td>\n",
       "      <td>Wsu|g13954.t1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5070</th>\n",
       "      <td>OG0005070</td>\n",
       "      <td>Bfl|BL07670</td>\n",
       "      <td>Bst|g3564.t1</td>\n",
       "      <td>Cmu|g21355.t1</td>\n",
       "      <td>Cpa|g28532.t1</td>\n",
       "      <td>Mme|g1681.t1</td>\n",
       "      <td>Wsu|g918.t1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5071</th>\n",
       "      <td>OG0005071</td>\n",
       "      <td>Bfl|BL24248</td>\n",
       "      <td>Bst|g3733.t1</td>\n",
       "      <td>Cmu|g12936.t1</td>\n",
       "      <td>Cpa|g9834.t1</td>\n",
       "      <td>Mme|g11780.t1</td>\n",
       "      <td>Wsu|g5758.t1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5072</th>\n",
       "      <td>OG0005072</td>\n",
       "      <td>Bfl|BL17600</td>\n",
       "      <td>Bst|g2214.t1</td>\n",
       "      <td>Cmu|g10198.t1</td>\n",
       "      <td>Cpa|g23724.t1</td>\n",
       "      <td>Mme|g2806.t1</td>\n",
       "      <td>Wsu|g2085.t2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5073</th>\n",
       "      <td>OG0005073</td>\n",
       "      <td>Bfl|BL05536</td>\n",
       "      <td>Bst|g6089.t1</td>\n",
       "      <td>Cmu|g7630.t1</td>\n",
       "      <td>Cpa|g12058.t1</td>\n",
       "      <td>Mme|g12008.t1</td>\n",
       "      <td>Wsu|g7389.t1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5074</th>\n",
       "      <td>OG0005074</td>\n",
       "      <td>Bfl|BL03184</td>\n",
       "      <td>Bst|g18614.t1</td>\n",
       "      <td>Cmu|g25496.t1</td>\n",
       "      <td>Cpa|g7673.t1</td>\n",
       "      <td>Mme|g13850.t1</td>\n",
       "      <td>Wsu|g16454.t1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1252 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Orthogroup          Bfl            Bst            Cmu            Cpa  \\\n",
       "3794  OG0003794  Bfl|BL24099   Bst|g3299.t1  Cmu|g20770.t1  Cpa|g29337.t1   \n",
       "3795  OG0003795  Bfl|BL02556  Bst|g18397.t1  Cmu|g17124.t1   Cpa|g9366.t1   \n",
       "3796  OG0003796  Bfl|BL13883  Bst|g12973.t2  Cmu|g25307.t1  Cpa|g23633.t1   \n",
       "3797  OG0003797  Bfl|BL16740  Bst|g16271.t1  Cmu|g19511.t1   Cpa|g6068.t1   \n",
       "3798  OG0003798  Bfl|BL14725  Bst|g11405.t1  Cmu|g18040.t1   Cpa|g2841.t1   \n",
       "...         ...          ...            ...            ...            ...   \n",
       "5070  OG0005070  Bfl|BL07670   Bst|g3564.t1  Cmu|g21355.t1  Cpa|g28532.t1   \n",
       "5071  OG0005071  Bfl|BL24248   Bst|g3733.t1  Cmu|g12936.t1   Cpa|g9834.t1   \n",
       "5072  OG0005072  Bfl|BL17600   Bst|g2214.t1  Cmu|g10198.t1  Cpa|g23724.t1   \n",
       "5073  OG0005073  Bfl|BL05536   Bst|g6089.t1   Cmu|g7630.t1  Cpa|g12058.t1   \n",
       "5074  OG0005074  Bfl|BL03184  Bst|g18614.t1  Cmu|g25496.t1   Cpa|g7673.t1   \n",
       "\n",
       "                Mme            Wsu  \n",
       "3794   Mme|g1166.t1    Wsu|g824.t1  \n",
       "3795  Mme|g13681.t1  Wsu|g16010.t1  \n",
       "3796  Mme|g25764.t2  Wsu|g20045.t1  \n",
       "3797  Mme|g18655.t1  Wsu|g11821.t1  \n",
       "3798  Mme|g16154.t1  Wsu|g13954.t1  \n",
       "...             ...            ...  \n",
       "5070   Mme|g1681.t1    Wsu|g918.t1  \n",
       "5071  Mme|g11780.t1   Wsu|g5758.t1  \n",
       "5072   Mme|g2806.t1   Wsu|g2085.t2  \n",
       "5073  Mme|g12008.t1   Wsu|g7389.t1  \n",
       "5074  Mme|g13850.t1  Wsu|g16454.t1  \n",
       "\n",
       "[1252 rows x 7 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get the single copy orthologues from the orthogroups file from orthofinder results. \n",
    "synteny.single_copy_orthologues() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "27e6d0de-c716-401e-88bb-eca3bc7e0f67",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Sc7u5tJ_1007</td>\n",
       "      <td>EVM</td>\n",
       "      <td>gene</td>\n",
       "      <td>847</td>\n",
       "      <td>3719</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"BL25399\"; copy_id \"BL25399_1\"; covera...</td>\n",
       "      <td>gene_id \"BL25399\"; copy_id \"BL25399_1\"; covera...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Sc7u5tJ_1014</td>\n",
       "      <td>EVM</td>\n",
       "      <td>gene</td>\n",
       "      <td>156</td>\n",
       "      <td>1489</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"BL02338\"; copy_id \"BL02338_1\"; covera...</td>\n",
       "      <td>gene_id \"BL02338\"; copy_id \"BL02338_1\"; covera...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Sc7u5tJ_1022</td>\n",
       "      <td>EVM</td>\n",
       "      <td>gene</td>\n",
       "      <td>3300</td>\n",
       "      <td>4921</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"BL13249\"; copy_id \"BL13249_1\"; covera...</td>\n",
       "      <td>gene_id \"BL13249\"; copy_id \"BL13249_1\"; covera...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Sc7u5tJ_1022</td>\n",
       "      <td>EVM</td>\n",
       "      <td>gene</td>\n",
       "      <td>7663</td>\n",
       "      <td>14871</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"BL13248\"; copy_id \"BL13248_1\"; covera...</td>\n",
       "      <td>gene_id \"BL13248\"; copy_id \"BL13248_1\"; covera...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Sc7u5tJ_1022</td>\n",
       "      <td>EVM</td>\n",
       "      <td>gene</td>\n",
       "      <td>15865</td>\n",
       "      <td>24273</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"BL13247\"; copy_id \"BL13247_1\"; covera...</td>\n",
       "      <td>gene_id \"BL13247\"; copy_id \"BL13247_1\"; covera...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28177</th>\n",
       "      <td>Sc7u5tJ_97</td>\n",
       "      <td>EVM</td>\n",
       "      <td>gene</td>\n",
       "      <td>9101</td>\n",
       "      <td>10795</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"BL22972\"; copy_id \"BL22972_1\"; covera...</td>\n",
       "      <td>gene_id \"BL22972\"; copy_id \"BL22972_1\"; covera...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28178</th>\n",
       "      <td>Sc7u5tJ_97</td>\n",
       "      <td>EVM</td>\n",
       "      <td>gene</td>\n",
       "      <td>11209</td>\n",
       "      <td>12522</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"BL22971\"; copy_id \"BL22971_1\"; covera...</td>\n",
       "      <td>gene_id \"BL22971\"; copy_id \"BL22971_1\"; covera...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28179</th>\n",
       "      <td>Sc7u5tJ_97</td>\n",
       "      <td>EVM</td>\n",
       "      <td>gene</td>\n",
       "      <td>18441</td>\n",
       "      <td>41685</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"BL22970\"; copy_id \"BL22970_1\"; covera...</td>\n",
       "      <td>gene_id \"BL22970\"; copy_id \"BL22970_1\"; covera...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28180</th>\n",
       "      <td>Sc7u5tJ_985</td>\n",
       "      <td>EVM</td>\n",
       "      <td>gene</td>\n",
       "      <td>155</td>\n",
       "      <td>2722</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"BL04168\"; copy_id \"BL04168_1\"; covera...</td>\n",
       "      <td>gene_id \"BL04168\"; copy_id \"BL04168_1\"; covera...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28181</th>\n",
       "      <td>Sc7u5tJ_996</td>\n",
       "      <td>EVM</td>\n",
       "      <td>gene</td>\n",
       "      <td>1883</td>\n",
       "      <td>2188</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"BL25405\"; copy_id \"BL25405_1\"; covera...</td>\n",
       "      <td>gene_id \"BL25405\"; copy_id \"BL25405_1\"; covera...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>28182 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           sequence source feature  start    end score strand frame  \\\n",
       "0      Sc7u5tJ_1007    EVM    gene    847   3719     .      +     .   \n",
       "1      Sc7u5tJ_1014    EVM    gene    156   1489     .      +     .   \n",
       "2      Sc7u5tJ_1022    EVM    gene   3300   4921     .      +     .   \n",
       "3      Sc7u5tJ_1022    EVM    gene   7663  14871     .      +     .   \n",
       "4      Sc7u5tJ_1022    EVM    gene  15865  24273     .      +     .   \n",
       "...             ...    ...     ...    ...    ...   ...    ...   ...   \n",
       "28177    Sc7u5tJ_97    EVM    gene   9101  10795     .      -     .   \n",
       "28178    Sc7u5tJ_97    EVM    gene  11209  12522     .      -     .   \n",
       "28179    Sc7u5tJ_97    EVM    gene  18441  41685     .      +     .   \n",
       "28180   Sc7u5tJ_985    EVM    gene    155   2722     .      -     .   \n",
       "28181   Sc7u5tJ_996    EVM    gene   1883   2188     .      -     .   \n",
       "\n",
       "                                               attribute  \\\n",
       "0      gene_id \"BL25399\"; copy_id \"BL25399_1\"; covera...   \n",
       "1      gene_id \"BL02338\"; copy_id \"BL02338_1\"; covera...   \n",
       "2      gene_id \"BL13249\"; copy_id \"BL13249_1\"; covera...   \n",
       "3      gene_id \"BL13248\"; copy_id \"BL13248_1\"; covera...   \n",
       "4      gene_id \"BL13247\"; copy_id \"BL13247_1\"; covera...   \n",
       "...                                                  ...   \n",
       "28177  gene_id \"BL22972\"; copy_id \"BL22972_1\"; covera...   \n",
       "28178  gene_id \"BL22971\"; copy_id \"BL22971_1\"; covera...   \n",
       "28179  gene_id \"BL22970\"; copy_id \"BL22970_1\"; covera...   \n",
       "28180  gene_id \"BL04168\"; copy_id \"BL04168_1\"; covera...   \n",
       "28181  gene_id \"BL25405\"; copy_id \"BL25405_1\"; covera...   \n",
       "\n",
       "                                              annotation  \n",
       "0      gene_id \"BL25399\"; copy_id \"BL25399_1\"; covera...  \n",
       "1      gene_id \"BL02338\"; copy_id \"BL02338_1\"; covera...  \n",
       "2      gene_id \"BL13249\"; copy_id \"BL13249_1\"; covera...  \n",
       "3      gene_id \"BL13248\"; copy_id \"BL13248_1\"; covera...  \n",
       "4      gene_id \"BL13247\"; copy_id \"BL13247_1\"; covera...  \n",
       "...                                                  ...  \n",
       "28177  gene_id \"BL22972\"; copy_id \"BL22972_1\"; covera...  \n",
       "28178  gene_id \"BL22971\"; copy_id \"BL22971_1\"; covera...  \n",
       "28179  gene_id \"BL22970\"; copy_id \"BL22970_1\"; covera...  \n",
       "28180  gene_id \"BL04168\"; copy_id \"BL04168_1\"; covera...  \n",
       "28181  gene_id \"BL25405\"; copy_id \"BL25405_1\"; covera...  \n",
       "\n",
       "[28182 rows x 10 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read the gtf file to a python dataframe. This may take a moment. \n",
    "# Use the 'feature' parameter to filter for only a type of row, ex. 'CDS'\n",
    "# Use the 'protein_id' parameter to search for protein_id in the annotation column\n",
    "# Use the 'equivalence' parameter to define an alternative separator (ex. '=' rather than ' ')\n",
    "\n",
    "# Tip: If you are uncertain about which annotation types or features to use, \n",
    "# run the line with no additional parameters first and examine the resulting dataframe. \n",
    "synteny.gtf_to_dataframe('Bfl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "7a886ecb-d018-4e0e-86c2-b51be540235d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CAUYWC010000002.1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>CDS</td>\n",
       "      <td>570969</td>\n",
       "      <td>570987</td>\n",
       "      <td>0.34</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>transcript_id \"g1.t1\"; gene_id \"g1\";</td>\n",
       "      <td>g1.t1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>CAUYWC010000002.1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>CDS</td>\n",
       "      <td>586843</td>\n",
       "      <td>586856</td>\n",
       "      <td>0.49</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"g2.t1\"; gene_id \"g2\";</td>\n",
       "      <td>g2.t1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>CAUYWC010000002.1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>CDS</td>\n",
       "      <td>799606</td>\n",
       "      <td>799660</td>\n",
       "      <td>0.99</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>transcript_id \"g3.t1\"; gene_id \"g3\";</td>\n",
       "      <td>g3.t1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>CAUYWC010000002.1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>CDS</td>\n",
       "      <td>805524</td>\n",
       "      <td>805529</td>\n",
       "      <td>0.87</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"g4.t1\"; gene_id \"g4\";</td>\n",
       "      <td>g4.t1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>CAUYWC010000002.1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>CDS</td>\n",
       "      <td>834389</td>\n",
       "      <td>834774</td>\n",
       "      <td>0.81</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>transcript_id \"g5.t1\"; gene_id \"g5\";</td>\n",
       "      <td>g5.t1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>744975</th>\n",
       "      <td>OY755168.1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>CDS</td>\n",
       "      <td>45752836</td>\n",
       "      <td>45752844</td>\n",
       "      <td>0.89</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"g21340.t1\"; gene_id \"g21340\";</td>\n",
       "      <td>g21340.t1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>744984</th>\n",
       "      <td>OY755168.1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>CDS</td>\n",
       "      <td>45880500</td>\n",
       "      <td>45880567</td>\n",
       "      <td>0.86</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>transcript_id \"g21341.t1\"; gene_id \"g21341\";</td>\n",
       "      <td>g21341.t1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>744996</th>\n",
       "      <td>OY755168.1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>CDS</td>\n",
       "      <td>45904430</td>\n",
       "      <td>45904930</td>\n",
       "      <td>0.68</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"g21342.t1\"; gene_id \"g21342\";</td>\n",
       "      <td>g21342.t1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>745002</th>\n",
       "      <td>OY755168.1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>CDS</td>\n",
       "      <td>45904990</td>\n",
       "      <td>45905624</td>\n",
       "      <td>0.14</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"g21343.t1\"; gene_id \"g21343\";</td>\n",
       "      <td>g21343.t1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>745014</th>\n",
       "      <td>OY755168.1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>CDS</td>\n",
       "      <td>45942263</td>\n",
       "      <td>45942338</td>\n",
       "      <td>0.58</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"g21344.t1\"; gene_id \"g21344\";</td>\n",
       "      <td>g21344.t1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25115 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 sequence    source feature     start       end score strand  \\\n",
       "3       CAUYWC010000002.1  AUGUSTUS     CDS    570969    570987  0.34      -   \n",
       "30      CAUYWC010000002.1  AUGUSTUS     CDS    586843    586856  0.49      +   \n",
       "42      CAUYWC010000002.1  AUGUSTUS     CDS    799606    799660  0.99      -   \n",
       "54      CAUYWC010000002.1  AUGUSTUS     CDS    805524    805529  0.87      -   \n",
       "78      CAUYWC010000002.1  AUGUSTUS     CDS    834389    834774  0.81      -   \n",
       "...                   ...       ...     ...       ...       ...   ...    ...   \n",
       "744975         OY755168.1  AUGUSTUS     CDS  45752836  45752844  0.89      +   \n",
       "744984         OY755168.1  AUGUSTUS     CDS  45880500  45880567  0.86      -   \n",
       "744996         OY755168.1  AUGUSTUS     CDS  45904430  45904930  0.68      +   \n",
       "745002         OY755168.1  AUGUSTUS     CDS  45904990  45905624  0.14      +   \n",
       "745014         OY755168.1  AUGUSTUS     CDS  45942263  45942338  0.58      +   \n",
       "\n",
       "       frame                                     attribute annotation  \n",
       "3          1          transcript_id \"g1.t1\"; gene_id \"g1\";      g1.t1  \n",
       "30         0          transcript_id \"g2.t1\"; gene_id \"g2\";      g2.t1  \n",
       "42         1          transcript_id \"g3.t1\"; gene_id \"g3\";      g3.t1  \n",
       "54         0          transcript_id \"g4.t1\"; gene_id \"g4\";      g4.t1  \n",
       "78         2          transcript_id \"g5.t1\"; gene_id \"g5\";      g5.t1  \n",
       "...      ...                                           ...        ...  \n",
       "744975     0  transcript_id \"g21340.t1\"; gene_id \"g21340\";  g21340.t1  \n",
       "744984     2  transcript_id \"g21341.t1\"; gene_id \"g21341\";  g21341.t1  \n",
       "744996     0  transcript_id \"g21342.t1\"; gene_id \"g21342\";  g21342.t1  \n",
       "745002     0  transcript_id \"g21343.t1\"; gene_id \"g21343\";  g21343.t1  \n",
       "745014     0  transcript_id \"g21344.t1\"; gene_id \"g21344\";  g21344.t1  \n",
       "\n",
       "[25115 rows x 10 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Making the dataframes: \n",
    "synteny.gtf_to_dataframe('Bfl', annotation_type = 'gene_id',)\n",
    "synteny.gtf_to_dataframe('Bst', annotation_type = 'transcript_id', feature = 'CDS')\n",
    "synteny.gtf_to_dataframe('Cmu', annotation_type = 'transcript_id', feature = 'CDS')\n",
    "synteny.gtf_to_dataframe('Cpa', annotation_type = 'transcript_id', feature = 'CDS')\n",
    "synteny.gtf_to_dataframe('Mme', annotation_type = 'transcript_id', feature = 'CDS')\n",
    "synteny.gtf_to_dataframe('Wsu', annotation_type = 'transcript_id', feature = 'CDS')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "1a3b0efc-ce60-456d-88a1-806bd2131fe1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3             Wsu|g1.t1\n",
       "30            Wsu|g2.t1\n",
       "42            Wsu|g3.t1\n",
       "54            Wsu|g4.t1\n",
       "78            Wsu|g5.t1\n",
       "              ...      \n",
       "744975    Wsu|g21340.t1\n",
       "744984    Wsu|g21341.t1\n",
       "744996    Wsu|g21342.t1\n",
       "745002    Wsu|g21343.t1\n",
       "745014    Wsu|g21344.t1\n",
       "Name: annotation, Length: 25115, dtype: object"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Modify the annotation names to include the species codes followed by a pipe\n",
    "# This matches the proteome_add_species() function. \n",
    "synteny.gtf_add_species()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "61684228-24f3-4cba-b220-1a401b8d845e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CAJZBW010000001.1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>CDS</td>\n",
       "      <td>532</td>\n",
       "      <td>660</td>\n",
       "      <td>0.68</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"g1.t1\"; gene_id \"g1\";</td>\n",
       "      <td>Mme|g1.t1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>CAJZBW010000001.1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1677</td>\n",
       "      <td>1883</td>\n",
       "      <td>0.68</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"g2.t1\"; gene_id \"g2\";</td>\n",
       "      <td>Mme|g2.t1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>CAJZBW010000001.1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>CDS</td>\n",
       "      <td>7458</td>\n",
       "      <td>7633</td>\n",
       "      <td>0.51</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"g3.t1\"; gene_id \"g3\";</td>\n",
       "      <td>Mme|g3.t1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>CAJZBW010000001.1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>CDS</td>\n",
       "      <td>9495</td>\n",
       "      <td>9749</td>\n",
       "      <td>0.94</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"g4.t1\"; gene_id \"g4\";</td>\n",
       "      <td>Mme|g4.t1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>CAJZBW010000001.1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>CDS</td>\n",
       "      <td>17803</td>\n",
       "      <td>17867</td>\n",
       "      <td>0.57</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"g5.t1\"; gene_id \"g5\";</td>\n",
       "      <td>Mme|g5.t1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1013191</th>\n",
       "      <td>OU612075.1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>CDS</td>\n",
       "      <td>23357678</td>\n",
       "      <td>23358242</td>\n",
       "      <td>0.99</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>transcript_id \"g25905.t1\"; gene_id \"g25905\";</td>\n",
       "      <td>Mme|g25905.t1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1013209</th>\n",
       "      <td>OU612075.1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>CDS</td>\n",
       "      <td>23362280</td>\n",
       "      <td>23362404</td>\n",
       "      <td>0.6</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"g25906.t1\"; gene_id \"g25906\";</td>\n",
       "      <td>Mme|g25906.t1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1013218</th>\n",
       "      <td>OU612075.1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>CDS</td>\n",
       "      <td>23364373</td>\n",
       "      <td>23364678</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"g25907.t1\"; gene_id \"g25907\";</td>\n",
       "      <td>Mme|g25907.t1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1013224</th>\n",
       "      <td>OU612075.1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>CDS</td>\n",
       "      <td>23364881</td>\n",
       "      <td>23365243</td>\n",
       "      <td>0.99</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"g25908.t1\"; gene_id \"g25908\";</td>\n",
       "      <td>Mme|g25908.t1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1013233</th>\n",
       "      <td>OU612075.1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>CDS</td>\n",
       "      <td>23366606</td>\n",
       "      <td>23367343</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"g25909.t1\"; gene_id \"g25909\";</td>\n",
       "      <td>Mme|g25909.t1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>31934 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  sequence    source feature     start       end score strand  \\\n",
       "3        CAJZBW010000001.1  AUGUSTUS     CDS       532       660  0.68      +   \n",
       "12       CAJZBW010000001.1  AUGUSTUS     CDS      1677      1883  0.68      +   \n",
       "24       CAJZBW010000001.1  AUGUSTUS     CDS      7458      7633  0.51      +   \n",
       "33       CAJZBW010000001.1  AUGUSTUS     CDS      9495      9749  0.94      +   \n",
       "39       CAJZBW010000001.1  AUGUSTUS     CDS     17803     17867  0.57      +   \n",
       "...                    ...       ...     ...       ...       ...   ...    ...   \n",
       "1013191         OU612075.1  AUGUSTUS     CDS  23357678  23358242  0.99      -   \n",
       "1013209         OU612075.1  AUGUSTUS     CDS  23362280  23362404   0.6      +   \n",
       "1013218         OU612075.1  AUGUSTUS     CDS  23364373  23364678     1      -   \n",
       "1013224         OU612075.1  AUGUSTUS     CDS  23364881  23365243  0.99      -   \n",
       "1013233         OU612075.1  AUGUSTUS     CDS  23366606  23367343     1      +   \n",
       "\n",
       "        frame                                     attribute     annotation  \n",
       "3           0          transcript_id \"g1.t1\"; gene_id \"g1\";      Mme|g1.t1  \n",
       "12          0          transcript_id \"g2.t1\"; gene_id \"g2\";      Mme|g2.t1  \n",
       "24          0          transcript_id \"g3.t1\"; gene_id \"g3\";      Mme|g3.t1  \n",
       "33          0          transcript_id \"g4.t1\"; gene_id \"g4\";      Mme|g4.t1  \n",
       "39          0          transcript_id \"g5.t1\"; gene_id \"g5\";      Mme|g5.t1  \n",
       "...       ...                                           ...            ...  \n",
       "1013191     1  transcript_id \"g25905.t1\"; gene_id \"g25905\";  Mme|g25905.t1  \n",
       "1013209     0  transcript_id \"g25906.t1\"; gene_id \"g25906\";  Mme|g25906.t1  \n",
       "1013218     0  transcript_id \"g25907.t1\"; gene_id \"g25907\";  Mme|g25907.t1  \n",
       "1013224     0  transcript_id \"g25908.t1\"; gene_id \"g25908\";  Mme|g25908.t1  \n",
       "1013233     0  transcript_id \"g25909.t1\"; gene_id \"g25909\";  Mme|g25909.t1  \n",
       "\n",
       "[31934 rows x 10 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "synteny.species_data['Mme']['gene_rows']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "509bd463-9164-480c-9fc7-c167c70863ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Looking at both the single copy orthogroups and the gene annotations, there will\n",
    "# sometimes be a slight 'mismatch' between the gene ids of these groups. For instance, \n",
    "# choosing transcript_id vs gene_id in the previous step would produce annotations with/without \n",
    "# a .t# suffix. While this discrepency will not significantly alter the results, it will\n",
    "# cause difficulty in matching orthologous genes to their annotations. \n",
    "# \n",
    "# If this is an issue and a species has an extra suffix in the proteome used initially, \n",
    "# we can use the truncate_sco('Species_code', 'Suffix') function. Alternatively, the \n",
    "# truncate_gene_rows() function does the same thing for gene row. \n",
    "#\n",
    "# This is unnecessary in the current example, but here is the syntax for how it would be run: \n",
    "# synteny.truncate_sco('Mme', '.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "4662d285-7592-4e76-ba2e-856e54a9157b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Wsu</th>\n",
       "      <th>sequence</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Wsu|g824.t1</td>\n",
       "      <td>OY755158.1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>CDS</td>\n",
       "      <td>10918680</td>\n",
       "      <td>10918690</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"g824.t1\"; gene_id \"g824\";</td>\n",
       "      <td>Wsu|g824.t1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Wsu|g16010.t1</td>\n",
       "      <td>OY755165.1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>CDS</td>\n",
       "      <td>34398365</td>\n",
       "      <td>34398379</td>\n",
       "      <td>0.82</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"g16010.t1\"; gene_id \"g16010\";</td>\n",
       "      <td>Wsu|g16010.t1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Wsu|g20045.t1</td>\n",
       "      <td>OY755167.1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>CDS</td>\n",
       "      <td>56807019</td>\n",
       "      <td>56807205</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"g20045.t1\"; gene_id \"g20045\";</td>\n",
       "      <td>Wsu|g20045.t1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Wsu|g11821.t1</td>\n",
       "      <td>OY755162.1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>CDS</td>\n",
       "      <td>36510878</td>\n",
       "      <td>36510979</td>\n",
       "      <td>0.99</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"g11821.t1\"; gene_id \"g11821\";</td>\n",
       "      <td>Wsu|g11821.t1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Wsu|g13954.t1</td>\n",
       "      <td>OY755164.1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>CDS</td>\n",
       "      <td>5598219</td>\n",
       "      <td>5598694</td>\n",
       "      <td>0.96</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>transcript_id \"g13954.t1\"; gene_id \"g13954\";</td>\n",
       "      <td>Wsu|g13954.t1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1247</th>\n",
       "      <td>Wsu|g918.t1</td>\n",
       "      <td>OY755158.1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>CDS</td>\n",
       "      <td>13057033</td>\n",
       "      <td>13057251</td>\n",
       "      <td>0.82</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"g918.t1\"; gene_id \"g918\";</td>\n",
       "      <td>Wsu|g918.t1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1248</th>\n",
       "      <td>Wsu|g5758.t1</td>\n",
       "      <td>OY755160.1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>CDS</td>\n",
       "      <td>8082399</td>\n",
       "      <td>8082441</td>\n",
       "      <td>0.66</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"g5758.t1\"; gene_id \"g5758\";</td>\n",
       "      <td>Wsu|g5758.t1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1249</th>\n",
       "      <td>Wsu|g2085.t2</td>\n",
       "      <td>OY755158.1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>CDS</td>\n",
       "      <td>41112718</td>\n",
       "      <td>41112780</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"g2085.t2\"; gene_id \"g2085\";</td>\n",
       "      <td>Wsu|g2085.t2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1250</th>\n",
       "      <td>Wsu|g7389.t1</td>\n",
       "      <td>OY755160.1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>CDS</td>\n",
       "      <td>64387045</td>\n",
       "      <td>64387471</td>\n",
       "      <td>0.77</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"g7389.t1\"; gene_id \"g7389\";</td>\n",
       "      <td>Wsu|g7389.t1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1251</th>\n",
       "      <td>Wsu|g16454.t1</td>\n",
       "      <td>OY755165.1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>CDS</td>\n",
       "      <td>50365391</td>\n",
       "      <td>50365540</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"g16454.t1\"; gene_id \"g16454\";</td>\n",
       "      <td>Wsu|g16454.t1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1252 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                Wsu    sequence    source feature     start       end score  \\\n",
       "0       Wsu|g824.t1  OY755158.1  AUGUSTUS     CDS  10918680  10918690     1   \n",
       "1     Wsu|g16010.t1  OY755165.1  AUGUSTUS     CDS  34398365  34398379  0.82   \n",
       "2     Wsu|g20045.t1  OY755167.1  AUGUSTUS     CDS  56807019  56807205     1   \n",
       "3     Wsu|g11821.t1  OY755162.1  AUGUSTUS     CDS  36510878  36510979  0.99   \n",
       "4     Wsu|g13954.t1  OY755164.1  AUGUSTUS     CDS   5598219   5598694  0.96   \n",
       "...             ...         ...       ...     ...       ...       ...   ...   \n",
       "1247    Wsu|g918.t1  OY755158.1  AUGUSTUS     CDS  13057033  13057251  0.82   \n",
       "1248   Wsu|g5758.t1  OY755160.1  AUGUSTUS     CDS   8082399   8082441  0.66   \n",
       "1249   Wsu|g2085.t2  OY755158.1  AUGUSTUS     CDS  41112718  41112780     1   \n",
       "1250   Wsu|g7389.t1  OY755160.1  AUGUSTUS     CDS  64387045  64387471  0.77   \n",
       "1251  Wsu|g16454.t1  OY755165.1  AUGUSTUS     CDS  50365391  50365540     1   \n",
       "\n",
       "     strand frame                                     attribute     annotation  \n",
       "0         +     0      transcript_id \"g824.t1\"; gene_id \"g824\";    Wsu|g824.t1  \n",
       "1         -     0  transcript_id \"g16010.t1\"; gene_id \"g16010\";  Wsu|g16010.t1  \n",
       "2         +     0  transcript_id \"g20045.t1\"; gene_id \"g20045\";  Wsu|g20045.t1  \n",
       "3         -     0  transcript_id \"g11821.t1\"; gene_id \"g11821\";  Wsu|g11821.t1  \n",
       "4         -     2  transcript_id \"g13954.t1\"; gene_id \"g13954\";  Wsu|g13954.t1  \n",
       "...     ...   ...                                           ...            ...  \n",
       "1247      -     0      transcript_id \"g918.t1\"; gene_id \"g918\";    Wsu|g918.t1  \n",
       "1248      +     0    transcript_id \"g5758.t1\"; gene_id \"g5758\";   Wsu|g5758.t1  \n",
       "1249      -     0    transcript_id \"g2085.t2\"; gene_id \"g2085\";   Wsu|g2085.t2  \n",
       "1250      +     0    transcript_id \"g7389.t1\"; gene_id \"g7389\";   Wsu|g7389.t1  \n",
       "1251      -     0  transcript_id \"g16454.t1\"; gene_id \"g16454\";  Wsu|g16454.t1  \n",
       "\n",
       "[1252 rows x 11 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Stitch together the orthogroup results with the associated gtf details to get the coordinates \n",
    "# for each gene. \n",
    "synteny.merge_gtf()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "4adc300c-2f39-4ca6-be72-832d1bdc7acf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>alg</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7948</td>\n",
       "      <td>Ea</td>\n",
       "      <td>Bfl|BL09372</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8096</td>\n",
       "      <td>H</td>\n",
       "      <td>Bfl|BL24099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9923</td>\n",
       "      <td>F</td>\n",
       "      <td>Bfl|BL02556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9315</td>\n",
       "      <td>N</td>\n",
       "      <td>Bfl|BL13883</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3413</td>\n",
       "      <td>C2</td>\n",
       "      <td>Bfl|BL16740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2356</th>\n",
       "      <td>3992</td>\n",
       "      <td>N</td>\n",
       "      <td>Bfl|BL26045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2357</th>\n",
       "      <td>1335</td>\n",
       "      <td>B1</td>\n",
       "      <td>Bfl|BL04267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2358</th>\n",
       "      <td>8287</td>\n",
       "      <td>D</td>\n",
       "      <td>Bfl|BL05536</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2359</th>\n",
       "      <td>12843</td>\n",
       "      <td>I</td>\n",
       "      <td>Bfl|BL21913</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2360</th>\n",
       "      <td>8564</td>\n",
       "      <td>F</td>\n",
       "      <td>Bfl|BL03184</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2361 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      index alg      gene_id\n",
       "0      7948  Ea  Bfl|BL09372\n",
       "1      8096   H  Bfl|BL24099\n",
       "2      9923   F  Bfl|BL02556\n",
       "3      9315   N  Bfl|BL13883\n",
       "4      3413  C2  Bfl|BL16740\n",
       "...     ...  ..          ...\n",
       "2356   3992   N  Bfl|BL26045\n",
       "2357   1335  B1  Bfl|BL04267\n",
       "2358   8287   D  Bfl|BL05536\n",
       "2359  12843   I  Bfl|BL21913\n",
       "2360   8564   F  Bfl|BL03184\n",
       "\n",
       "[2361 rows x 3 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Pass in the species code corresponding to the proteome with the ALG information. \n",
    "# The default parameters assume ids of the format index_alg_gene (ex. 9197_A1b_BL14725), \n",
    "# such that columns = ['index', 'alg', 'gene_id'] and delimiter = '_', however these can \n",
    "# all be changed as long as 'gene_id' and 'alg' are included. \n",
    "# 'mapping_add_species' is set to True by default and adds the species code followed by a | \n",
    "# prior to each id, but this option can also be changed through the intake parameters. \n",
    "synteny.parse_algs('Bfl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "ff94b9d1-e803-4d53-9b59-148ff99859e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "synteny.trace_algs('Bfl') "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "13959937-e816-409e-89d5-e5a05440f984",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Create a 'trace' group based on which chromosome each gene appears on for a given species. \n",
    "# # This allows us to later colour genes using this grouping. \n",
    "# synteny.trace_chromosomes('Mme') "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "1a357bd9-8b49-4de8-8885-72dba715370d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Complete</th>\n",
       "      <th>sequence</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>TRACE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Complete</td>\n",
       "      <td>OY755158.1</td>\n",
       "      <td>10918680</td>\n",
       "      <td>10918690</td>\n",
       "      <td>H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Complete</td>\n",
       "      <td>OY755165.1</td>\n",
       "      <td>34398365</td>\n",
       "      <td>34398379</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Complete</td>\n",
       "      <td>OY755167.1</td>\n",
       "      <td>56807019</td>\n",
       "      <td>56807205</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Complete</td>\n",
       "      <td>OY755162.1</td>\n",
       "      <td>36510878</td>\n",
       "      <td>36510979</td>\n",
       "      <td>C2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Complete</td>\n",
       "      <td>OY755164.1</td>\n",
       "      <td>5598219</td>\n",
       "      <td>5598694</td>\n",
       "      <td>A1b</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1247</th>\n",
       "      <td>Complete</td>\n",
       "      <td>OY755158.1</td>\n",
       "      <td>13057033</td>\n",
       "      <td>13057251</td>\n",
       "      <td>J2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1248</th>\n",
       "      <td>Complete</td>\n",
       "      <td>OY755160.1</td>\n",
       "      <td>8082399</td>\n",
       "      <td>8082441</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1249</th>\n",
       "      <td>Complete</td>\n",
       "      <td>OY755158.1</td>\n",
       "      <td>41112718</td>\n",
       "      <td>41112780</td>\n",
       "      <td>I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1250</th>\n",
       "      <td>Complete</td>\n",
       "      <td>OY755160.1</td>\n",
       "      <td>64387045</td>\n",
       "      <td>64387471</td>\n",
       "      <td>D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1251</th>\n",
       "      <td>Complete</td>\n",
       "      <td>OY755165.1</td>\n",
       "      <td>50365391</td>\n",
       "      <td>50365540</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1252 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Complete    sequence     start       end TRACE\n",
       "0     Complete  OY755158.1  10918680  10918690     H\n",
       "1     Complete  OY755165.1  34398365  34398379     F\n",
       "2     Complete  OY755167.1  56807019  56807205     N\n",
       "3     Complete  OY755162.1  36510878  36510979    C2\n",
       "4     Complete  OY755164.1   5598219   5598694   A1b\n",
       "...        ...         ...       ...       ...   ...\n",
       "1247  Complete  OY755158.1  13057033  13057251    J2\n",
       "1248  Complete  OY755160.1   8082399   8082441     G\n",
       "1249  Complete  OY755158.1  41112718  41112780     I\n",
       "1250  Complete  OY755160.1  64387045  64387471     D\n",
       "1251  Complete  OY755165.1  50365391  50365540     F\n",
       "\n",
       "[1252 rows x 5 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Starting from the second column, provide a list to define the columns of the output file. \n",
    "# If a given item in the list is a column of the orthogene_coords dataframe, the information \n",
    "# from that column will be used. Otherwise, the string will be repeated for the entire column. \n",
    "# The first column will be an index. \n",
    "\n",
    "# If the keyword 'TRACE' is used, the alg or chromosome trace will be included in the column. \n",
    "columns = ['Complete', 'sequence', 'start', 'end', 'TRACE'] \n",
    "synteny.clean_coords(columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "171243c7-cee7-478b-8af1-0e08af7706bf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Bfl</th>\n",
       "      <th>sequence</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Bfl|BL24099</td>\n",
       "      <td>Sc7u5tJ_1552</td>\n",
       "      <td>EVM</td>\n",
       "      <td>gene</td>\n",
       "      <td>20259441</td>\n",
       "      <td>20295001</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"BL24099\"; copy_id \"BL24099_1\"; covera...</td>\n",
       "      <td>Bfl|BL24099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Bfl|BL02556</td>\n",
       "      <td>Sc7u5tJ_1559</td>\n",
       "      <td>EVM</td>\n",
       "      <td>gene</td>\n",
       "      <td>2023441</td>\n",
       "      <td>2049604</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"BL02556\"; copy_id \"BL02556_1\"; covera...</td>\n",
       "      <td>Bfl|BL02556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Bfl|BL13883</td>\n",
       "      <td>Sc7u5tJ_566</td>\n",
       "      <td>EVM</td>\n",
       "      <td>gene</td>\n",
       "      <td>21977322</td>\n",
       "      <td>22002718</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"BL13883\"; copy_id \"BL13883_1\"; covera...</td>\n",
       "      <td>Bfl|BL13883</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Bfl|BL16740</td>\n",
       "      <td>Sc7u5tJ_366</td>\n",
       "      <td>EVM</td>\n",
       "      <td>gene</td>\n",
       "      <td>2633405</td>\n",
       "      <td>2639698</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"BL16740\"; copy_id \"BL16740_1\"; covera...</td>\n",
       "      <td>Bfl|BL16740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Bfl|BL14725</td>\n",
       "      <td>Sc7u5tJ_339</td>\n",
       "      <td>EVM</td>\n",
       "      <td>gene</td>\n",
       "      <td>24190631</td>\n",
       "      <td>24209724</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"BL14725\"; copy_id \"BL14725_1\"; covera...</td>\n",
       "      <td>Bfl|BL14725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1247</th>\n",
       "      <td>Bfl|BL07670</td>\n",
       "      <td>Sc7u5tJ_1590</td>\n",
       "      <td>EVM</td>\n",
       "      <td>gene</td>\n",
       "      <td>24518767</td>\n",
       "      <td>24523958</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"BL07670\"; copy_id \"BL07670_1\"; covera...</td>\n",
       "      <td>Bfl|BL07670</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1248</th>\n",
       "      <td>Bfl|BL24248</td>\n",
       "      <td>Sc7u5tJ_1565</td>\n",
       "      <td>EVM</td>\n",
       "      <td>gene</td>\n",
       "      <td>10986510</td>\n",
       "      <td>11024484</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"BL24248\"; copy_id \"BL24248_1\"; covera...</td>\n",
       "      <td>Bfl|BL24248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1249</th>\n",
       "      <td>Bfl|BL17600</td>\n",
       "      <td>Sc7u5tJ_1568</td>\n",
       "      <td>EVM</td>\n",
       "      <td>gene</td>\n",
       "      <td>33518158</td>\n",
       "      <td>33536473</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"BL17600\"; copy_id \"BL17600_1\"; covera...</td>\n",
       "      <td>Bfl|BL17600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1250</th>\n",
       "      <td>Bfl|BL05536</td>\n",
       "      <td>Sc7u5tJ_1571</td>\n",
       "      <td>EVM</td>\n",
       "      <td>gene</td>\n",
       "      <td>4845665</td>\n",
       "      <td>4846864</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"BL05536\"; copy_id \"BL05536_1\"; covera...</td>\n",
       "      <td>Bfl|BL05536</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1251</th>\n",
       "      <td>Bfl|BL03184</td>\n",
       "      <td>Sc7u5tJ_1559</td>\n",
       "      <td>EVM</td>\n",
       "      <td>gene</td>\n",
       "      <td>12856618</td>\n",
       "      <td>12870837</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"BL03184\"; copy_id \"BL03184_1\"; covera...</td>\n",
       "      <td>Bfl|BL03184</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1252 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Bfl      sequence source feature     start       end score  \\\n",
       "0     Bfl|BL24099  Sc7u5tJ_1552    EVM    gene  20259441  20295001     .   \n",
       "1     Bfl|BL02556  Sc7u5tJ_1559    EVM    gene   2023441   2049604     .   \n",
       "2     Bfl|BL13883   Sc7u5tJ_566    EVM    gene  21977322  22002718     .   \n",
       "3     Bfl|BL16740   Sc7u5tJ_366    EVM    gene   2633405   2639698     .   \n",
       "4     Bfl|BL14725   Sc7u5tJ_339    EVM    gene  24190631  24209724     .   \n",
       "...           ...           ...    ...     ...       ...       ...   ...   \n",
       "1247  Bfl|BL07670  Sc7u5tJ_1590    EVM    gene  24518767  24523958     .   \n",
       "1248  Bfl|BL24248  Sc7u5tJ_1565    EVM    gene  10986510  11024484     .   \n",
       "1249  Bfl|BL17600  Sc7u5tJ_1568    EVM    gene  33518158  33536473     .   \n",
       "1250  Bfl|BL05536  Sc7u5tJ_1571    EVM    gene   4845665   4846864     .   \n",
       "1251  Bfl|BL03184  Sc7u5tJ_1559    EVM    gene  12856618  12870837     .   \n",
       "\n",
       "     strand frame                                          attribute  \\\n",
       "0         -     .  gene_id \"BL24099\"; copy_id \"BL24099_1\"; covera...   \n",
       "1         +     .  gene_id \"BL02556\"; copy_id \"BL02556_1\"; covera...   \n",
       "2         +     .  gene_id \"BL13883\"; copy_id \"BL13883_1\"; covera...   \n",
       "3         +     .  gene_id \"BL16740\"; copy_id \"BL16740_1\"; covera...   \n",
       "4         -     .  gene_id \"BL14725\"; copy_id \"BL14725_1\"; covera...   \n",
       "...     ...   ...                                                ...   \n",
       "1247      +     .  gene_id \"BL07670\"; copy_id \"BL07670_1\"; covera...   \n",
       "1248      -     .  gene_id \"BL24248\"; copy_id \"BL24248_1\"; covera...   \n",
       "1249      -     .  gene_id \"BL17600\"; copy_id \"BL17600_1\"; covera...   \n",
       "1250      +     .  gene_id \"BL05536\"; copy_id \"BL05536_1\"; covera...   \n",
       "1251      -     .  gene_id \"BL03184\"; copy_id \"BL03184_1\"; covera...   \n",
       "\n",
       "       annotation  \n",
       "0     Bfl|BL24099  \n",
       "1     Bfl|BL02556  \n",
       "2     Bfl|BL13883  \n",
       "3     Bfl|BL16740  \n",
       "4     Bfl|BL14725  \n",
       "...           ...  \n",
       "1247  Bfl|BL07670  \n",
       "1248  Bfl|BL24248  \n",
       "1249  Bfl|BL17600  \n",
       "1250  Bfl|BL05536  \n",
       "1251  Bfl|BL03184  \n",
       "\n",
       "[1252 rows x 11 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "synteny.species_data['Bfl']['orthogene_coords']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "0627e72b-6c1f-4990-bc3b-7e40f08648a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write the coordinate files to the output folder. \n",
    "synteny.write_coords()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1fd1eaf6-07aa-4bb4-938d-5f79d5cf2e4a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d49e3d92-6b8a-4fc4-926d-545cba7d849f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36630081-13af-4e99-a5a8-0e5857bbed04",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50eaf21c-e54e-4599-84c8-d4600109b7cb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3c39b69-9694-4147-954a-aebeaba9c5a6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
